internal/worker/run.go

```diff
@@ -191,7 +211,89 @@ func (w *Worker) runStep(ctx context.Context, log logr.Logger, sandbox cri.CRI,
 	log.Info(aurora.Green("Step completed").String())
 }
 
-func (w *Worker) runTaskLogic(ctx context.Context, syslog, log logr.Logger, slot int, claim *tcqueue.TaskClaim, wr io.Writer) error {
+func (w *Worker) uploadArtifact(
+	ctx context.Context, log logr.Logger, claim *tcqueue.TaskClaim,
+	rootContainer cri.Container, artifact *config.Artifact,
+	deps []<-chan error, ch chan<- error,
+) {
+	var err error
+	defer func() {
+		ch <- err
+		close(ch)
+	}()
+
+	// Wait for all dependencies.
+	for _, ch := range deps {
+		err = <-ch
+		if err != nil {
+			return
+		}
+	}
+
+	if artifact.Type != config.FileArtifact {
+		err = fmt.Errorf("only file artifacts are supported")
+		return
+	}
+
+	srcPath := artifact.Path
+	if !path.IsAbs(srcPath) {
+		srcPath = path.Join("/workspace", artifact.Path)
+	}
+
+	var r io.ReadCloser
+	r, info, err := rootContainer.ReadFiles(ctx, srcPath)
```

Comment:
does this immediately shadow `var r` and `err`, meaning any error here won't actually be captured by the defer?

Reply:
a possible solution is instead of having this function do the logic and write to the error channel, move the logic to a private function (or inner/lambda function?) that returns an error, so you use normal error handling flows inside the function, and the outer function can just report the error as needed.

Reply:
> does this immediately shadow `var r` and `err`, meaning any error here won't actually be captured by the defer?

No, it doesn't shadow, it just assigns to the existing variables. The `var r` sets the type of that variable.

Reply:
my badddddddddddd